automower_ble/protocol.py

```diff
@@ -0,0 +1,504 @@
+
+        if request_length > 0:
+            self.request_data += request_data
+
+        self.request_data[2] = len(self.request_data) - 2  # Length
+
+        self.request_data[9] = crc(self.request_data, 1, 8)# CRC
+
+        # Two last bytes are crc and 0x03
+        self.request_data.append(crc(self.request_data, 1, len(self.request_data) - 1))
+        self.request_data.append(0x03)# Hard coded value
+
+        return self.request_data
+
+    def parse_response(self, response_data: bytearray) -> int | None:
+        response_length = response_data[17]
+        data = response_data[19:19 + response_length]
+        response = dict()
+        dpos = 0 # data position
+        for name, dtype in self.response_data_type.items(): 
+            if (dtype == "no_response"):
+                return None
+            elif (dtype == "tUnixTime") or (dtype == "uint32"):
+                response[name]=int.from_bytes(data[dpos : dpos + 4], byteorder="little")
+                dpos += 4
+            elif dtype == "uint16":
+                response[name]=int.from_bytes(data[dpos : dpos + 2], byteorder="little")
+                dpos += 2
+            elif (dtype == "uint8") or (dtype == "bool"):
+                response[name] = data[dpos]
+                dpos += 1
+            else:
+                raise ValueError("Unknown data type: " + dtype)
+        if dpos != len(data):
+            raise ValueError("Data length mismatch. Read %d bytes of %d" % (dpos, len(data)))
+
+        return response
+
+    def validate_response(self, response_data: bytearray) -> bool:
+        if response_data[0] != 0x02:
+            return False
+
+        if response_data[1] != 0xFD:
+            return False
+
+        if response_data[3] != 0x00:
+            return False
+
+        id = self.channel_id.to_bytes(4, byteorder="little")
+        if response_data[4] != id[0]:
+            return False
+        if response_data[5] != id[1]:
+            return False
+        if response_data[6] != id[2]:
+            return False
+        if response_data[7] != id[3]:
+            return False
+
+        if response_data[8] != 0x01:
+            # This is a valid config, but we don't support it
+            # return m1656b(decodeState, c10786f);
+            return False
+
+        if response_data[9] != crc(response_data, 1, 8):
+            return False
+
+        if response_data[10] != 0x01:
+            return False
+
+        if response_data[11] != 0xAF:
+            return False
+
+        major_bytes = self.major.to_bytes(4, byteorder="little")
+        if response_data[12] != major_bytes[0]:
+            return False
+        if response_data[13] != major_bytes[1]:
+            return False
+        if response_data[14] != self.minor:
+            return False
+
+        if response_data[15] != 0x00:
+            return False
+
+        if response_data[16] != 0x00:
+            return False
+
+        return True
+
+class BLEClient:
```

Comment:
This feels like it should be in `mower.py` instead of `protocol.py`

Reply:
I tried to move the code related to protocol and BLE communication to `protocol.py` in order to keep the Mower class smaller just representing and handling the state of the mower. 

Let me know if this makes sense or if you'd like me to move (or rename) BLEClient? Or possibly revert those changes and keep everything in a big Mower class?